File: 200-number-of-islands/200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */

const dfs = function (grid, i, j) {
	if (
		i < 0 ||
		j < 0 ||
		i > grid.length - 1 ||
		j > grid[0].length - 1 ||
		grid[i][j] === '0'
	) {
		return 0;
	}
	grid[i][j] = '0';
	dfs(grid, i - 1, j);
	dfs(grid, i + 1, j);
	dfs(grid, i, j - 1);
	dfs(grid, i, j + 1);
};

var numIslands = function (grid) {
	let islands = 0;

	const m = grid.length;
	grid.forEach((row, i) => {
		row.forEach((col, j) => {
			if (grid[i][j] === '1') {
				islands++;
				dfs(grid, i, j);
			}
		});
	});

	return islands;
};

File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    grid: list[list[str]]
    m: int
    n: int

    def numIslands(self, grid: list[list[str]]) -> int:
        self.grid = grid
        self.m = len(grid)
        self.n = len(grid[0])
        cnt = 0

        for i in range(0, self.m):
            for j in range(0, self.n):
                if grid[i][j] == "1":
                    cnt += 1
                    self.dfs(i,j)
        return cnt

    def dfs(self, i, j):
        if i < 0 or i >= self.m or j < 0 or j >= self.n or self.grid[i][j] == "0":
            return

        self.grid[i][j] = "0"
        self.dfs(i, j+1)
        self.dfs(i+1, j)
        self.dfs(i, j-1)
        self.dfs(i-1, j)